koku/api/report/queries.py

```diff
@@ -0,0 +1,331 @@
+                # get last 10 days
+                end = this_hour
+                start = ReportQueryHandler.n_days_ago(this_hour, 10)
+            else:
+                # get last 30 days
+                end = this_hour
+                start = ReportQueryHandler.n_days_ago(this_hour, 30)
+
+        if start and end and resolution == 'daily':
+            interval = ReportQueryHandler.list_days(start, end)
+        if start and end and resolution == 'monthly':
+            interval = ReportQueryHandler.list_months(start, end)
+        return (start, end, interval)
+
+    def _add_interval_data(self, interval, query_data):
+        """Add the interval data if not present in cost usage data.
+
+        Args:
+            interval    (List[DateTime]): Valid interval list sliced by resolution
+            query_data  (List(Dict)): Queried data
+        Returns:
+            (List[Dict]): List of data for all intervals sliced by resolution
+
+        """
+        current = {}
+        data = []
+        resolution = self.get_resolution()
+        if query_data:
+            current = query_data.pop()
+        while interval:
+            interval_item = interval.pop().date()
+            if resolution == 'daily':
+                interval_str = interval_item.strftime('%Y-%m-%d')
+            else:
+                interval_str = interval_item.strftime('%Y-%m')
+            cur_date = current.get('date')
+            if isinstance(cur_date, datetime.datetime):
+                cur_date = cur_date.date()
+            if cur_date == interval_item:
+                data.append({'date': interval_str,
+                             'total_cost': current.get('total_cost'),
+                             'units': 'USD'})
+                if query_data:
+                    current = query_data.pop()
+            else:
+                data.append({'date': interval_str,
+                             'total_cost': None,
+                             'units': None})
+        return data
+
+    def _format_query_response(self):
+        """Format the query response with data.
+
+        Returns:
+            (Dict): Dictionary response of query params, data, and total
+
+        """
+        output = copy.deepcopy(self.query_parameters)
+        output['data'] = self.query_data
+        output['total'] = self.query_sum
+        return output
+
+    def execute_query(self):
+        """Execute query and return provided data.
+
+        Returns:
+            (Dict): Dictionary response of query params, data, and total
+
+        """
+        query_sum = None
+        data = []
+        if self.get_resolution() == 'daily':
+            trunc_func = TruncDay
+        else:
+            trunc_func = TruncMonth
+
+        with tenant_context(self.tenant):
+            start, end, days = self._get_time_frame_filter()
+            time_frame_query = AWSCostEntryLineItem.objects.filter(
+                usage_start__gte=start,
+                usage_end__lte=end)
```

Comment:
Not to be changed now, but just to keep in our back pocket if this query acts slowly with lots of data... The line item has a foreign key to the cost entry table, and an index on that foreign key. One thing to try later would be doing a query on the cost entry table to get the date ranges, then filter the line item table on that range of ids.

Reply:
ok.